electionleaflets/apps/leaflets/forms.py

```diff
@@ -63,26 +46,94 @@ class Meta:
-            obj.name,
-            party_name,
-        )
+    def create_option(self, name, value, label, selected, index, subindex=None,
+                      attrs=None):
+        if not label:
+            label = "Not Listed"
+        else:
+            label = u"{0} ({1})".format(
+                label["person"]["name"],
+                label["party"]["party_name"],
+            )
+        return super(PeopleRadioWidget, self).create_option(name, value, label, selected, index,
+                                                            subindex, attrs)
 
 
 class PeopleForm(forms.Form):
+    # A temporary top-X list of parties per register to show if the candidate
+    # isn't shown. We will eventually create this list based on real data.
+    HARDCODED_PARTIES = {
+        'gb': [
+            (52, 'Conservative and Unionist Party'),
+            (63, 'Green Party'),
+            (53, 'Labour Party'),
+            (90, 'Liberal Democrats'),
+            (102, 'Scottish National Party (SNP)'),
+            (7931, 'The Brexit Party'),
+            (77, 'Plaid Cymru - The Party of Wales'),
+        ],
+        'ni': [
+            (103, 'Alliance - Alliance Party of Northern Ireland'),
+            (70, 'Democratic Unionist Party - D.U.P.'),
+            (55, 'SDLP (Social Democratic & Labour Party)'),
+            (39, 'Sinn Féin'),
+            (83, 'Ulster Unionist Party'),
+        ]
+    }
+
     def __init__(self, *args, **kwargs):
         super(PeopleForm, self).__init__(*args, **kwargs)
-        if 'people' in kwargs['initial']:
+        signer = Signer()
+
+        if "postcode_results" in kwargs['initial']:
+
+            postcode_results = kwargs['initial']["postcode_results"].json()
+
+            # We have a response, parse each candidacy in to a set
+            # of unique people as people can stand in more than one ballot
+            # for a postcode
+            unique_people = OrderedDict()
+            for date in postcode_results["dates"]:
+                for ballot in date["ballots"]:
+                    for candidacy in ballot["candidates"]:
+                        # Until we have better live lookup of YNR in EL, we'll
+                        # embed the results in our form fields. Not ideal. We'll
+                        # sign the data so people can't inject garbage into the
+                        # database.
+                        data = {
+                            "ynr_party_id": candidacy["party"]["party_id"],
+                            "ynr_party_name": candidacy["party"]["party_name"],
+                            "ynr_person_id": candidacy["person"]["ynr_id"],
+                            "ynr_person_name": candidacy["person"]["name"],
+                            "ballot_id": ballot['ballot_paper_id'],
+                        }
+                        person_key = signer.sign(json.dumps(data))
+                        unique_people[person_key] = candidacy
```

Comment:
If you flatten the person list here but you encode the `ballot_id` into the `person_key`, if the same person stands in >1 ballot that person's name will occur twice in the flat list but it will be completely unclear to the user why that's happening. It also means that depending on which of the 2 instances of the same person the user picks, we'll tag the leaflet with a different ballot id, but as a user you don't know that the 2 labels with the same person name on them do different things. If you're going to flatten it I think you've got to abandon the concept of a ballot as a useful data point here.

Reply:
True, and soon we'll group by ballot. But this is a short term fix for the GE. At the moment the data is completely out of date.

Reply:
See the screenshot further up for how we'll solve this later